spark/src/main/scala/org/apache/spark/sql/delta/UpdateExpressionsSupport.scala

```diff
@@ -131,27 +114,29 @@ trait UpdateExpressionsSupport extends CastSupport with SQLConfHelper with Analy
                   ArrayType(toEt, containsNull = true)
                 )
             }
-          case NestedMapConversion(from: MapType, to: MapType) if !Cast.canCast(from, to) =>
-            // Manually convert map values if the types are not compatible to allow schema
+          case (from: MapType, to: MapType) if !Cast.canCast(from, to) =>
+            // Manually convert map keys and values if the types are not compatible to allow schema
             // evolution. This is slower than direct cast so we only do it when required.
-            if (!from.keyType.sameType(to.keyType)) {
-              throw DeltaErrors.unsupportedMapKeyTypeChange(
-                mapType = from,
-                fromKeyType = from.keyType,
-                toKeyType = to.keyType
-              )
-            }
-            val valueConverter: (Expression, Expression) => Expression = (key, _) =>
-              castIfNeeded(GetMapValue(fromExpression, key), to.valueType, allowStructEvolution)
-            val transformLambdaFunc = {
+            def createMapConverter(convert: (Expression, Expression) => Expression): Expression = {
               val keyVar = NamedLambdaVariable("keyVar", from.keyType, nullable = false)
-              val valueVar = NamedLambdaVariable("valueVar", from.valueType, from.valueContainsNull)
-              LambdaFunction(valueConverter(keyVar, valueVar), Seq(keyVar, valueVar))
+              val valueVar =
+                NamedLambdaVariable("valueVar", from.valueType, from.valueContainsNull)
+              LambdaFunction(convert(keyVar, valueVar), Seq(keyVar, valueVar))
+            }
+
+            var transformedKeysAndValues = fromExpression
+            if (from.keyType != to.keyType) {
+              transformedKeysAndValues = TransformKeys(fromExpression, createMapConverter {
+                (key, _) => castIfNeeded(key, to.keyType, allowStructEvolution)
+              })
+            }
+
+            if (from.valueType != to.valueType) {
+              transformedKeysAndValues = TransformValues(fromExpression, createMapConverter {
```

Comment:
Doesn't this break if we need to transform both keys and values, as we are using `fromExpression`, instead of `transformedKeysAndValues`? We probably require another test for this

Reply:
Ouch, yes that slipped in unintentionally. I fixed it and added the missing test.